### PR TITLE
Partially implement define queries

### PR DESCRIPTION
**Pull request by krishnangovindraj**
## Usage and product changes
Implement basic functionality of `define` queries.

## Implementation
We iterate over the parsed `definable`s and apply changes in the following order:
`types ≤ type annotations < sub < value-type < relates < plays ≤ owns`

Some functionality remains:
* Implement overrides,
* Figure out how to elegantly handle idempotence and conflicting declarations in a define query. e.g.
```
define
relation rel, relate rol[];
relation subrel relates subrol as rol;
```

### CODE REVIEW

File: query/define.rs [hunk 1]
```diff
@@ -6,107 +6,606 @@
...
+    snapshot: &mut impl WritableSnapshot,
+    type_manager: &TypeManager,
+    definables: &[Definable],
+) -> Result<(), DefineError> {
+    filter_variants!(Definable::Function : definables)
+        .try_for_each(|declaration| define_functions(snapshot, type_manager, definables))?;
+    Ok(())
+}
+
+fn define_struct<'a>(
```
Comment (dmitrii-ubskii) [inline query/define.rs:138]: Unused lifetime?

File: query/define.rs [hunk 2]
```diff
@@ -6,107 +6,606 @@
...
-        if let Definable::Struct(struct_definable) = definable {
-            let name = struct_definable.ident.as_str();
-            let struct_key = type_manager.create_struct(snapshot, name.to_owned()).map_err(|err| {
-                DefineError::StructCreateError { source: err, struct_declaration: struct_definable.clone() }
-            })?;
+    &type_declaration
+        .capabilities
+        .iter()
+        .filter_map(|capability| try_unwrap!(CapabilityBase::Alias = &capability.base))
+        .try_for_each(|_| Err(DefineError::Unimplemented))?;
```
Comment (dmitrii-ubskii) [inline query/define.rs:274]: `filter_variants!()`

File: query/define.rs [hunk 3]
```diff
@@ -6,107 +6,606 @@
...
 
+macro_rules! unwrap_or_else {
+    ($variant:path = $to_unwrap:expr ; $else_:block) => {
+        if let $variant(inner) = $to_unwrap {
+            inner
+        } else {
+            $else_
+        };
+    };
+}
```
Comment (dmitrii-ubskii) [inline query/define.rs:55]: This should be deleted and the usages replaced with `let-else`:
```rs
let $variant(inner) = $to_unwrap else { return $else_; };
```
or
```rs
let $variant(inner) = $to_unwrap else { continue; };
```
(the inside of the `else {}` must diverge, i.e. `return`, `panic!`, `break`, or `continue`)

**Review by dmitrii-ubskii — APPROVED** [review]
Overall approved, with a few notes to address

File: dependencies/vaticle/repositories.bzl [hunk 4]
```diff
@@ -42,6 +42,6 @@ def vaticle_typedb_protocol():
 def vaticle_typedb_behaviour():
     git_repository(
         name = "vaticle_typedb_behaviour",
-        remote = "https://github.com/typedb/typedb-behaviour",
-        commit = "ba6f933ddd26b77b6754f5e955c5eaee15dbf523", # sync-marker: do not remove this comment, this is used for sync-dependencies by @vaticle_typedb_behaviour
+        remote = "https://github.com/krishnangovindraj/typedb-behaviour",
+        commit = "8a48ba2547889d3ccf40c03a22a1be0aa39b91d0", # sync-marker: do not remove this comment, this is used for sync-dependencies by @vaticle_typedb_behaviour
```
Comment (flyingsilverfin) [inline dependencies/vaticle/repositories.bzl:46]: PR as well

File: ir/translation/tokens.rs [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    Kind as TypeQLKind,
+    ValueType as TypeQLValueType,
+};
+use concept::type_::annotation::{Annotation, AnnotationAbstract, AnnotationCardinality, AnnotationCascade, AnnotationDistinct, AnnotationIndependent, AnnotationKey, AnnotationRange, AnnotationRegex, AnnotationUnique, AnnotationValues};
+use encoding::graph::type_::Kind;
+use encoding::value::value_type::ValueType;
+use crate::LiteralParseError;
+use crate::translation::literal::translate_literal;
+
+pub fn translate_annotation(typeql_kind: &TypeQLAnnotation) -> Result<Annotation, LiteralParseError> {
```
Comment (flyingsilverfin) [inline ir/translation/tokens.rs:19]: nice good place for this stuff

File: query/util.rs
[the remark this answers is not in the document]
Comment (flyingsilverfin) [reply query/util.rs:1]: I dont mind keeping this for now until we see what usage patterns emerge across all the Query implementations

**Review by flyingsilverfin — APPROVED** [review]
Approved pending few comments

**Review by krishnangovindraj — COMMENTED** [review]
replies

File: encoding/graph/type_/mod.rs [hunk 6]
```diff
@@ -12,7 +12,7 @@ pub mod property;
 pub mod vertex;
 pub mod vertex_generator;
 
-#[derive(Copy, Clone)]
+#[derive(Copy, Clone, Ord, PartialOrd, Eq, PartialEq)]
```
[the remark this answers is not in the document]
Comment (krishnangovindraj) [reply encoding/graph/type_/mod.rs:15]: Seems unused. Will remove.

**Pull request merged**